cs/src/io/unsafe/InputPtrBuffer.cs

```diff
@@ -0,0 +1,269 @@
+        public byte ReadUInt8()
+        {
+            if (position >= end)
+            {
+                EndOfStream(1);
+            }
+            return buffer[position++];
+        }
+
+        /// <summary>
+        /// Read little-endian encoded 16-bit unsigned integer
+        /// </summary>
+        /// <exception cref="EndOfStreamException"/>
+        public ushort ReadUInt16()
+        {
+            if (position > end - sizeof(ushort))
+            {
+                EndOfStream(sizeof(ushort));
+            }
+            uint result = buffer[position++];
+            result |= ((uint)buffer[position++]) << 8;
+            return (ushort)result;
+        }
+
+        /// <summary>
+        /// Read little-endian encoded 32-bit unsigned integer
+        /// </summary>
+        /// <exception cref="EndOfStreamException"/>
+        public uint ReadUInt32()
+        {
+            if (position > end - sizeof(uint))
+            {
+                EndOfStream(sizeof(uint));
+            }
+            uint result = buffer[position++];
+            result |= ((uint)buffer[position++]) << 8;
+            result |= ((uint)buffer[position++]) << 16;
+            result |= ((uint)buffer[position++]) << 24;
+            return result;
+        }
+
+        /// <summary>
+        /// Read little-endian encoded 64-bit unsigned integer
+        /// </summary>
+        /// <exception cref="EndOfStreamException"/>
+        public ulong ReadUInt64()
+        {
+            if (position > end - sizeof(ulong))
+            {
+                EndOfStream(sizeof(ulong));
+            }
+
+            var result = *((ulong*)(buffer + position));
+            position += sizeof(ulong);
+            return result;
+        }
+
+        /// <summary>
+        /// Read little-endian encoded single precision ‎IEEE 754 float
+        /// </summary>
+        /// <exception cref="EndOfStreamException"/>
+        public float ReadFloat()
+        {
+            if (position > end - sizeof(float))
+            {
+                EndOfStream(sizeof(float));
+            }
+
+            var result = *((float*)(buffer + position));
+            position += sizeof(float);
+            return result;
+        }
+
+        /// <summary>
+        /// Read little-endian encoded double precision ‎IEEE 754 float
+        /// </summary>
+        /// <exception cref="EndOfStreamException"/>
+        public double ReadDouble()
+        {
+            if (position > end - sizeof(double))
+            {
+                EndOfStream(sizeof(double));
+            }
+            var result = *((double*)(buffer + position));
+            position += sizeof(double);
+            return result;
+        }
+
+        /// <summary>
+        /// Read an array of bytes verbatim
+        /// </summary>
+        /// <param name="count">Number of bytes to read</param>
+        /// <exception cref="EndOfStreamException"/>
+        public ArraySegment<byte> ReadBytes(int count)
```

Comment:
Yes, this is sub-optimal, making `byte*` slower than `byte[]` in this case. But this isn't that simple to fix because this is dictated by the type used to represent `blob` in generated classes and it percolates throughout, not just `IInputStream` but also protocol interfaces. And I'm not sure if it would even be desirable to allow representing `blob` as a `byte*` to the source buffer, lifetime management would be very error prone.